nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/validators/SNMPValidator.java

```diff
@@ -0,0 +1,100 @@
+
+    private final TargetConfiguration targetConfiguration;
+    private final List<ValidationResult> problems;
+
+    public SNMPValidator(final TargetConfiguration targetConfiguration, final List<ValidationResult> problems) {
+        this.targetConfiguration = targetConfiguration;
+        this.problems = problems;
+    }
+
+    public Collection<ValidationResult> validate() {
+        final boolean isVersion3 = SNMPVersion.SNMP_V3.getSnmpVersionDisplay().equals(targetConfiguration.getVersion());
+        final boolean isSecurityNameInvalid = isInvalid(targetConfiguration.getSecurityName());
+        final boolean isCommunityStringInvalid = isInvalid(targetConfiguration.getCommunityString());
+
+        if (isVersion3 && isSecurityNameInvalid) {
+            problems.add(new ValidationResult.Builder()
+                    .input("SNMP Security Name")
+                    .valid(false)
+                    .explanation("SNMP Security Name must be set with SNMPv3.")
+                    .build());
+            checkSecurityLevel(targetConfiguration, problems);
+
+        } else if (isCommunityStringInvalid) {
+            problems.add(new ValidationResult.Builder()
+                    .input("SNMP Community")
+                    .valid(false)
+                    .explanation("SNMP Community must be set with SNMPv1 and SNMPv2c.")
+                    .build());
+        }
+        return problems;
+    }
+
+    private void checkSecurityLevel(TargetConfiguration targetConfiguration, List<ValidationResult> problems) {
+
+        final boolean isAuthProtocolInvalid = isInvalid(targetConfiguration.getAuthProtocol());
+        final boolean isAuthPasswordInvalid = isInvalid(targetConfiguration.getAuthPassword());
+        final boolean isPrivacyProtocolInvalid = isInvalid(targetConfiguration.getPrivacyProtocol());
+        final boolean isPrivacyPasswordInvalid = isInvalid(targetConfiguration.getPrivacyPassword());
+        final SecurityLevel securityLevel = SecurityLevel.valueOf(targetConfiguration.getSecurityLevel());
+
+        if (isAuthNoPrivSecurityLevelInvalid(securityLevel, isAuthProtocolInvalid, isAuthPasswordInvalid)) {
+            problems.add(new ValidationResult.Builder()
+                    .input("SNMP Security Level")
+                    .valid(false)
+                    .explanation("Authentication protocol and password must be set when using authNoPriv security level.")
+                    .build());
+        }
+
+        if (isAuthPrivSecurityLevelInvalid(securityLevel, isAuthProtocolInvalid, isAuthPasswordInvalid, isPrivacyProtocolInvalid, isPrivacyPasswordInvalid)) {
+            problems.add(new ValidationResult.Builder()
+                    .input("SNMP Security Level")
+                    .valid(false)
+                    .explanation("All protocols and passwords must be set when using authPriv security level.")
+                    .build());
+        }
+    }
+
+    private boolean isInvalid(String property) {
```

Comment:
I think the use of `dependsOn()` would seriously simplify this validation, even so it would end up short enough to move it back to the `customValidation` method. Please take a look on `MergeContent`, it has some good example.